cpp/include/cugraph/edge_properties.hpp

```diff
@@ -0,0 +1,210 @@
+                                                 rmm::device_uvector<float>,
+                                                 rmm::device_uvector<double>,
+                                                 rmm::device_uvector<int32_t>,
+                                                 rmm::device_uvector<int64_t>,
+                                                 rmm::device_uvector<size_t>>;
+using numeric_device_span_t       = std::variant<std::monostate,
+                                                 raft::device_span<float>,
+                                                 raft::device_span<double>,
+                                                 raft::device_span<int32_t>,
+                                                 raft::device_span<int64_t>,
+                                                 raft::device_span<size_t>>;
+using const_numeric_device_span_t = std::variant<std::monostate,
+                                                 raft::device_span<float const>,
+                                                 raft::device_span<double const>,
+                                                 raft::device_span<int32_t const>,
+                                                 raft::device_span<int64_t const>,
+                                                 raft::device_span<size_t const>>;
+
+template <typename edge_t>
+using edge_property_views_t = std::variant<std::monostate,
+                                           cugraph::edge_property_view_t<edge_t, float const*>,
+                                           cugraph::edge_property_view_t<edge_t, double const*>,
+                                           cugraph::edge_property_view_t<edge_t, int32_t const*>,
+                                           cugraph::edge_property_view_t<edge_t, int64_t const*>,
+                                           cugraph::edge_property_view_t<edge_t, size_t const*>>;
+
+template <typename func_t>
+auto variant_type_dispatch(numeric_device_uvector_t& property, func_t func)
+{
+  if (std::holds_alternative<rmm::device_uvector<float>>(property)) {
+    auto& prop = std::get<rmm::device_uvector<float>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<rmm::device_uvector<double>>(property)) {
+    auto& prop = std::get<rmm::device_uvector<double>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<rmm::device_uvector<int32_t>>(property)) {
+    auto& prop = std::get<rmm::device_uvector<int32_t>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<rmm::device_uvector<int64_t>>(property)) {
+    auto& prop = std::get<rmm::device_uvector<int64_t>>(property);
+    return func(prop);
+  } else if (std::holds_alternative<rmm::device_uvector<size_t>>(property)) {
+    auto& prop = std::get<rmm::device_uvector<size_t>>(property);
+    return func(prop);
+  }
+
+  CUGRAPH_FAIL("unsupported variant type -- shouldn't happen");
```

Comment:
You don't get a warning here? (something related to missing return statement).

Better restructure this code as

```
else {
  CUGRAPH_EXPECTS(std::hold_alternative<rmm::device_uvector<size_t>>(property), "Invalid input argument: unsupported variant type.");
  ...;
}
```

Reply:
I don't see a warning here... but your suggestion avoids that ambiguity, I'll add it to the next push.